msdk-featuredetection/msdk-targeteddetection/src/main/java/io/github/msdk/featuredetection/targeteddetection/TargetedDetectionMethod.java

```diff
@@ -164,14 +173,25 @@ public TargetedDetectionMethod(@Nonnull List<IonAnnotation> ionAnnotations,
                 chromatogram.setDataPoints(newDataPoints);
             }
 
+            // Set the m/z value for the chromatogram
+            double[] mzValues = buildingChromatogram.getMzValues();
+            double sum = 0;
+            int values = 0;
+            for(int i=0; i < buildingChromatogram.getSize() ; i++) {
+                if (mzValues[i] > 0) {
+                    values++;
+                    sum = sum + mzValues[i];
+                }
+            }
+            double newMz = sum/values;
+            chromatogram.setMz(newMz);
+
             // Add the chromatogram to the chromatogram list

```

Comment:
Hi,
I think that both approaches make sense in different situations.
Generally speaking, a median based approach will be less sensitive
towards outliers.
Depending on the instrument's detector, its precision, and the method
used for binning/pre-processing
of the m/z values used for the single-ion chromatogram, the median
approach will result in an m/z value
that represents the larger number of m/z's in the sample, while the mean
approach will be distorted as soon
as binning has placed m/z's into adjacent bins or has moved them to the
current bin. You can make the mean more
robust by omitting the bottom and top k % of m/z values after sorting. A
related approach for preprocessing is taken
by XCMS/Camera, if I recall correctly.

To evaluate this for data from different vendors and different
preprocessing strategies, a first step would be
to observe the ratio of median and mean. Additionally, there are usually
characteristic distributions of m/z's for
specific instruments, although they may even vary over the detector's
range, so it is a classic tradeoff between
effort required and observable effect on the data quality.

Nils

Am 15.10.15 um 15:16 schrieb Tomáš Pluskal:
>
> In
> msdk-featuredetection/msdk-targeteddetection/src/main/java/io/github/msdk/featuredetection/targeteddetection/TargetedDetectionMethod.java
> <https://github.com/msdk/msdk/pull/19#discussion_r42122201>:
>
> > @@ -164,14 +173,25 @@ public TargetedDetectionMethod(@Nonnull List<IonAnnotation> ionAnnotations,
> >                  chromatogram.setDataPoints(newDataPoints);
> >              }
> >  
> > +            // Set the m/z value for the chromatogram
> > +            double[] mzValues = buildingChromatogram.getMzValues();
> > +            double sum = 0;
> > +            int values = 0;
> > +            for(int i=0; i < buildingChromatogram.getSize() ; i++) {
> > +                if (mzValues[i] > 0) {
> > +                    values++;
> > +                    sum = sum + mzValues[i];
> > +                }
> > +            }
> > +            double newMz = sum/values;
> > +            chromatogram.setMz(newMz);
> > +
> >              // Add the chromatogram to the chromatogram list
>
> That could be true. But it would be good to make a bit of evaluation
> before doing such change. Also, try discussing this with Matej - I
> think the original median concept probably came from him.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/msdk/msdk/pull/19/files#r42122201>.
>